sdk/python/kfp/components/placeholders.py

```diff
@@ -402,150 +186,76 @@ def container_with_if_placeholder(output_path: OutputPath(str),
-        struct_body = struct['IfPresent']
-
-        then = struct_body['Then']
-        then = [
-            maybe_convert_placeholder_string_to_placeholder(item)
-            for item in then
-        ] if isinstance(then, list) else then
-
-        else_ = struct_body.get('Else')
-        else_ = [
-            maybe_convert_placeholder_string_to_placeholder(item)
-            for item in else_
-        ] if isinstance(else_, list) else else_
-        kwargs = {
-            'input_name': struct_body['InputName'],
-            'then': then,
-            'else_': else_
+    def __init__(
+        self,
+        input_name: str,
+        then: Union['CommandLineElement', List['CommandLineElement']],
+        else_: Optional[Union['CommandLineElement',
+                              List['CommandLineElement']]] = None,
+    ) -> None:
+        self.input_name = input_name
+        self.then = then
+        self.else_ = else_
+
+    def _to_dict(self) -> Dict[str, Any]:
+        struct = {
+            'IfPresent': {
+                'InputName':
+                    self.input_name,
+                'Then': [
+                    convert_command_line_element_to_string_or_struct(e)
+                    for e in self.then
+                ] if isinstance(self.then, list) else
+                        convert_command_line_element_to_string_or_struct(
+                            self.then)
+            }
         }
-        return IfPresentPlaceholder(**kwargs)
-
-    def _transform_else(self) -> None:
-        """Use None instead of empty list for optional."""
-        self.else_ = None if self.else_ == [] else self.else_
-
-
-class CustomizedDecoder(json.JSONDecoder):
-
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        def parse_array(*_args, **_kwargs):
-            values, end = JSONArray(*_args, **_kwargs)
-            for i, item in enumerate(values):
-                if isinstance(item, dict):
-                    values[i] = json.dumps(item)
-            return values, end
+        if self.else_:
+            struct['IfPresent']['Else'] = [
+                convert_command_line_element_to_string_or_struct(e)
+                for e in self.else_
+            ] if isinstance(
+                self.else_,
+                list) else convert_command_line_element_to_string_or_struct(
+                    self.else_)
+        return struct
 
-        self.parse_array = parse_array
-        self.scan_once = py_make_scanner(self)
+    def _to_string(self) -> str:
+        return json.dumps(self._to_dict())
 
 
-def json_load_nested_placeholder_aware(
-    placeholder_string: str
-) -> Union[str, Dict[str, Union[str, List[str], dict]]]:
-    try:
-        return json.loads(placeholder_string, cls=CustomizedDecoder)
-    except json.JSONDecodeError:
-        return placeholder_string
+CONTAINER_PLACEHOLDERS = (IfPresentPlaceholder, ConcatPlaceholder)
```

Comment:
nit: prefix with underscore since you don't reference them outside the module. 

https://google.github.io/styleguide/pyguide.html#3162-naming-conventions

Reply:
Updated